### ISSUE TITLE
x/pkgsite/cmd/pkgsite: Examples not executable if pkg.name != basename(pkg.path)

### ISSUE THREAD

**Issue by alandonovan** [golang/go#12794]
Update: see https://github.com/golang/go/issues/12794#issuecomment-3667366323

% export GOPATH=$(pwd)
% cat src/foo1/foo.go
package foo

func Foo() {
}

% cat src/foo1/foo_test.go
package foo_test

import "foo1" // defines foo

func ExampleFoo() {
        foo.Foo()
}

% godoc -play -http :9999 &
% open http://localhost:9999/pkg/foo1/#pkg-examples

The example is shown but not executable (grey not yellow); renaming foo1 to foo makes it executable, as does using an explicit (redundant) renaming import:
   import foo "foo1"

Seems like some code in godoc is assuming pkg.name == basename(pkg.path) instead of finding the actual package name.

**Comment by rsc** [comment]
Not such a terrible assumption. :-)

**Comment by ysmolsky** [comment]
I see that this example is discarded as not executable in this block of code: https://github.com/golang/go/blob/master/src/go/doc/example.go#L211-L243:
```
	// Use unresolved identifiers to determine the imports used by this
	// example. The heuristic assumes package names match base import
	// paths for imports w/o renames (should be good enough most of the time).
	namedImports := make(map[string]string) // [name]path
	var blankImports []ast.Spec             // _ imports
	for _, s := range file.Imports {
		p, err := strconv.Unquote(s.Path.Value)
		if err != nil {
			continue
		}
		n := path.Base(p)
		if s.Name != nil {
			n = s.Name.Name
			switch n {
			case "_":
				blankImports = append(blankImports, s)
				continue
			case ".":
				// We can't resolve dot imports (yet).
				return nil
			}
		}
		if unresolved[n] {
			namedImports[n] = p
			delete(unresolved, n)
		}
	}
	// If there are other unresolved identifiers, give up because this
	// synthesized file is not going to build.
	if len(unresolved) > 0 {
		return nil
	}
```
Specifically, `unresolved=map[foo:true]` at line 242.

**Comment by ysmolsky** [comment]
@alandonovan can you hint me how I can figure out what is the actual package name is being imported given it's package name and path? Should program try to import the actual package to get its name?

**Comment by seankhliao** [comment]
With the above CL, it should work for most code formatted with goimports.
goimports adds package name aliases if it doesn't fit one of the conventional patterns (exact match, `/vN` major version, `go-` prefix).

Perhaps we can close this as good enough for most cases? cc @adonovan

**Comment by adonovan** [comment]
The original bug report was against x/tools/cmd/godoc, which is since been deleted. I just tested it against `go doc -http`, which runs a local pkgsite server, and it has the same problem, though it's possible that's because the explicit version of pkgsite that cmd/go requests has not yet picked up the fix of [CL 728280](https://go.dev/cl/728280). So I test pkgsite@master, and it still has the problem: that is, an explicit renaming import causes the Example's Run button to appear, but without it, the Example is not runnable.

(BTW: the example also needs an `// Output:` comment to be runnable, which I forgot to mention in the issue body.)

```
$ go version
go version go1.26-devel_ad91f5d241 Wed Dec 17 13:38:56 2025 -0800 darwin/arm64

$ nl -ba  foo1/foo.go 
     1	package foo
     2	
     3	func Foo() {
     4	}

$ nl -ba  foo1/foo_test.go 
     1	package foo_test
     2	
     3	import "foo1"
     4	
     5	func ExampleFoo() {
     6		foo.Foo()
     7		// Output:
     8		// foo
     9	}

$ pkgsite &
$ open http://localhost:8080/pkg/foo1#example-Foo
```

Observe there is no Executable Run button; then edit foo.go to `import foo "foo"` and reload, and observe the Run button.

cc: @matloob

**Comment by seankhliao** [comment]
right, [CL 728280](https://go.dev/cl/728280) is still only lexical, it doesn't actually try to pull in the dependencies to check their package names.

but it may be quite infrequent to have such mismatched imports anyway?

**Comment by adonovan** [comment]
> but it may be quite infrequent to have such mismatched imports anyway?

It may be infrequent, but it's still a bug, one that affected the late x/tools/cmd/godoc and still affects the current master branch of pkgsite built with the toolchain at master.

**Comment by pjweinb** [comment]
Closing as obsolete: originally filed against godoc -play in GOPATH mode. pkgsite is maintained in golang/pkgsite.
